File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void dfs(vector<int>& nums, vector<int>& container, vector<vector<int>>& answer, vector<bool>& visited){
        if(container.size() >= nums.size()){
            answer.push_back(container);
            return; 
        }

        for(int i = 0; i < nums.size(); i++){
            if(!visited[i]){
                visited[i] = true;
                container.push_back(nums[i]); 
                dfs(nums,container,answer,visited); 
                container.pop_back(); 
                visited[i] = false;  
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> answer; 
        vector<int> container; 
        vector<bool> visited(nums.size(), false); 
        dfs(nums,container,answer, visited); 
        return answer; 
    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    boolean[] visited = new boolean[11]; 
    public void dfs(List<List<Integer>> answer, List<Integer> container, int[] nums){
        if(container.size() >= nums.length){
            answer.add(new ArrayList<>(container)); 
            return; 
        }

        for(int i = 0; i < nums.length; i++){
            if(!visited[i]){
                visited[i] = true; 
                container.add(nums[i]); 
                dfs(answer, container, nums);
                visited[i] = false; 
                container.removeLast(); 
            }
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> answer = new ArrayList<>(); 
        List<Integer> container = new ArrayList<>(); 

        dfs(answer, container, nums); 

        return answer; 
    }
}